include/revng/Pipeline/Pipe.h

```diff
@@ -0,0 +1,311 @@
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Pipe, F, Containers, ArgsNames, Indexes);
+}
+
+class PipeWrapperBaseImpl {
+public:
+  virtual void run(Context &Ctx, ContainerSet &Containers) = 0;
+  virtual ContainersStatus
+  getRequirements(const ContainersStatus &Target) const = 0;
+  virtual ContainersStatus deduceResults(ContainersStatus &Target) const = 0;
+  virtual ~PipeWrapperBaseImpl() = default;
+  virtual std::unique_ptr<PipeWrapperBaseImpl> clone() const = 0;
+  virtual std::vector<std::string> getRunningContainersNames() const = 0;
+  virtual std::string getName() const = 0;
+  virtual void dump(std::ostream &OS, size_t indents) const = 0;
+};
+
+template<typename T>
+concept Printable = requires(T a) {
+  { a.print(dbg, 0) };
+};
+
+template<typename PipeType>
+class PipeWrapperImpl : public PipeWrapperBaseImpl {
+private:
+  PipeType Pipe;
+  std::vector<std::string> RunningContainersNames;
+
+public:
+  PipeWrapperImpl(PipeType Pipe,
+                  std::vector<std::string> RunningContainersNames) :
+    Pipe(std::move(Pipe)),
+    RunningContainersNames(std::move(RunningContainersNames)) {}
+
+  void run(Context &Ctx, ContainerSet &Containers) override {
+    invokePipeFunction(Ctx,
+                       Pipe,
+                       &PipeType::run,
+                       Containers,
+                       RunningContainersNames);
+  }
+
+  ContainersStatus
+  getRequirements(const ContainersStatus &Target) const override {
+    const auto &Contracts = Pipe.getContract();
+    auto ToReturn = Target;
+    for (const auto &Contract : llvm::reverse(Contracts))
+      ToReturn = Contract.deduceRequirements(ToReturn, RunningContainersNames);
+    return ToReturn;
+  }
+
+  ContainersStatus deduceResults(ContainersStatus &Target) const override {
+    const auto &Contracts = Pipe.getContract();
+    for (const auto &Contract : Contracts)
+      Contract.deduceResults(Target, RunningContainersNames);
+    return Target;
+  }
+
+  ~PipeWrapperImpl() override = default;
+
+  std::unique_ptr<PipeWrapperBaseImpl> clone() const override {
+    return std::make_unique<PipeWrapperImpl>(*this);
+  }
+
+  std::vector<std::string> getRunningContainersNames() const override {
+    return RunningContainersNames;
+  }
+
+  std::string getName() const override { return PipeType::Name; }
+
+  void dump(std::ostream &OS, size_t Indents) const override {
+    if constexpr (Printable<PipeType>)
+      Pipe.dump(OS, Indents);
+  }
+};
+
+} // namespace detail
+
+/// A pipe must be type erased somehow to become compatible with a
+/// pipeline, a PipeWrapper takes care of this issue, it can be
+/// constructed from any pipeline type, and it will expose the
+/// contract and run method of that enforcer.
+class PipeWrapper {
+private:
+  std::unique_ptr<detail::PipeWrapperBaseImpl> Pipe;
```

Comment:
Can we
```c++
using PipeWrapper = std::unique_ptr<detail::PipeWrapperBaseImpl> Pipe;
```
?

Or at least make this a `unique_ptr`-like (i.e., implement `*`/`->`), otherwise we have three classes that define the same set of methods with little benefit.

Reply:
not sure what you mean? we need a layer with the interface declaration, one with the implementation that overrides the interface, and one that hides the pointer. There is each method 3 times but the user only sees one.

Reply:
`PipeWrapper` is not much different than a plain `std::unique_ptr<detail::PipeWrapperBaseImpl>` but it forces you to write the list of methods one extra time, while you could just give back to users `PipeWrapperBaseImpl *` which is `virtual` and good to use directly.

Reply:
i tried dropping the wrapper, what happens is that multiple templates need to get out of the name space detail, exposing part of the implementation. 

Everywhere there was the creation of a pipeline now there is a unique_ptr that make the code less readable.
as an example, functions that are already not trivial such as 
```cpp
   template<typename... LLVMPipePassess>
      static PipeWrapper
      wrapLLVMPassess(std::string LLVMModuleName, LLVMPipePassess &&...P) {
        return PipeWrapper(LLVMPipeBase<Derived>(std::move(P)...),
                           { std::move(LLVMModuleName) });
      }
```
become a mess 
```cpp
  template<typename... LLVMPipePassess>
      static std::unique_ptr<PipeWrapperBase>
      wrapLLVMPassess(std::string LLVMModuleName, LLVMPipePassess &&...P) {
          using WrapperType = PipeWrapperImpl<LLVMPipeBase<Derived>>;
        return std::make_unsigned<WrapperType>(LLVMPipeBase<Derived>(std::move(P)...),
                           { std::move(LLVMModuleName) });
      }

```

Reply:
having the duplication of code there reaaally reduces the complexity in the users

Reply:
i will put a comment explaing what is going on instead if that is ok

Reply:
If it's just a matter of this is instantiated, make an helper function.

Otherwise, if you really think that class is beneficial, do not replicate the list of methods but offer `*`/`->` and expose the wrapped object. Basically make it behave more like a `unique_ptr`.

Reply:
ok, implemented as a pointer like